pkg/vsphere/vsan/dom_cache.go

```diff
@@ -0,0 +1,226 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package vsan is a workaround for vsan DOM object leaking issue caused by FileManager.DeleteDatastoreFile, see github issue #3787 and bugzilla issue #1808703
+// This file used draft vSphere API, which is subject to change in the future, so this workaround should be removed as soon as the DOM leaking issue is fixed by vSAN.
+package vsan
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/vic/pkg/errors"
+	"github.com/vmware/vic/pkg/trace"
+)
+
+const (
+	vsanNamePattern = "vsan:[a-zA-Z0-9-]*/"
+	vmdkFileSuffix  = ".vmdk"
+)
+
+var (
+	vsanNameRegex, _ = regexp.Compile(vsanNamePattern)
+)
+
+// vsanDatastoreCache is the map from vmdk file path to vsan dom object uuid, all vmdk files in the datastore will be cached
+type vsanDSDomCache struct {
+	ds   *object.Datastore
+	hvis *HostVsanInternalSystem
+	// vsan dom object uuid: vsan dom object path, this path does not have vsan namespace prefix in this format /vmfs/volumes/vsan:52932941b44e2147-f1490d38c9730c6d/.
+	// uuids keeps all vsan dom objects in this datastore, to improve VMDK uuid query performance, because API does not provide any way to filter based on object path
+	uuids map[string]string
+	// vsan dom object path:object uuid. This path has same truncation with the above map
+	paths map[string]string
+	m     *sync.Mutex
+}
+
+// DeleteVMDKDoms deletes vmdk dom objects if the vmdk file exists in dom cache, if not, return undeleted files
+func (v *vsanDSDomCache) DeleteVMDKDoms(ctx context.Context, paths []string) ([]string, error) {
+	defer trace.End(trace.Begin(fmt.Sprintf("paths: %s", paths)))
+	var uuids []string
+	var ret []string
+
+	v.m.Lock()
+	for _, path := range paths {
+		if !strings.HasSuffix(path, vmdkFileSuffix) {
+			log.Debugf("non vmdk file %s, no need to delete here", path)
+			continue
+		}
+		if uuid, ok := v.paths[path]; ok {
+			uuids = append(uuids, uuid)
+		} else {
+			ret = append(ret, path)
+		}
+	}
+	v.m.Unlock()
+
+	force := true
+	res, err := v.hvis.DeleteVsanObjects(ctx, uuids, &force)
+	var deletedDoms []string
+	if err != nil {
+		return ret, DomDeleteError{
+			Err:         err,
+			FailedUuids: uuids,
+		}
+	}
+	var failedIds []string
+	for _, r := range res {
+		if !r.Success {
+			failedIds = append(failedIds, r.Uuid)
+		} else {
+			deletedDoms = append(deletedDoms, r.Uuid)
+		}
+	}
+	v.cleanDeletedUUIDs(deletedDoms)
+
+	if len(failedIds) > 0 {
+		return ret, DomDeleteError{
+			FailedUuids: failedIds,
+			Result:      res,
+		}
+	}
+	return ret, nil
+}
+
+func (v *vsanDSDomCache) cleanDeletedUUIDs(deletedDoms []string) {
+	defer trace.End(trace.Begin(fmt.Sprintf("%s", deletedDoms)))
+
+	var deletedVMDKs []string
+	v.m.Lock()
```

Comment:
I don't want to lock vsphere operations, but to make it clear, removed the internal method with lock/unlock